### PR TITLE
feat(artifacts): Enable GCE deploy stage to deploy an artifact

### PR DESCRIPTION
**Pull request by ezimanyi**
The deploy stage for the GCE provider now gives the option of deploying a specified upstream artifact rather than deploying the result of the most recent Bake or Find Artifact stage.

### CODE REVIEW

**Comment by ezimanyi** [comment]
The changes here depend on the Orca changes in https://github.com/spinnaker/orca/pull/2049 to properly work; the change to Orca should be merged first.

File: app/scripts/modules/core/src/artifact/expectedArtifactSelector.component.ts [hunk 1]
```diff
@@ -23,21 +23,20 @@ class ExpectedArtifactSelectorComponent implements IComponentOptions {
...
+                         class="form-control input-sm" required>
                 <ui-select-match>{{ $select.selected | summarizeExpectedArtifact }}</ui-select-match>
                 <ui-select-choices repeat="expected.id as expected in ctrl.expectedArtifacts">
                   <span>{{ expected | summarizeExpectedArtifact }}</span>
                 </ui-select-choices>
               </ui-select>
             </div>
           </div>
-          <div class="form-group">
+          <div ng-if="ctrl.account-field" class="form-group">
```
Comment (sbwsg) [inline app/scripts/modules/core/src/artifact/expectedArtifactSelector.component.ts]: should this be `accountField` to match the public prop name on line 10?

File: app/scripts/modules/google/src/help/gce.help.ts [hunk 2]
```diff
@@ -13,6 +13,16 @@ const helpContents: {[key: string]: string} = {
       and can be configured within the server group creation dialogue under <b>Port Name Mapping</b>.`,
   'gce.httpLoadBalancer.pathRule.paths': 'For example, <b>/path</b> in <b>example.com/path</b>',
   'gce.httpLoadBalancer.port': 'HTTP requests can be load balanced based on port 80 or port 8080. HTTPS requests can be load balanced on port 443.',
+  'gce.image.artifact': 'The artifact that is to be deployed to this cluster.  The artifact should represent a deployable image.',
+  'gce.image.source': `
+      <p>Where the image to deploy is read from.</p>
+            <p>
```
Comment (sbwsg) [inline app/scripts/modules/google/src/help/gce.help.ts]: Indent looks wrong here

File: app/scripts/modules/core/src/artifact/imageSourceSelector.component.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
+/*
```
Comment (danielpeach) [inline app/scripts/modules/core/src/artifact/imageSourceSelector.component.ts]: Please omit this. The build leaves comments in the generated app (or at least it did in the past), which we try to keep as small as possible.

File: app/scripts/modules/core/src/artifact/imageSourceSelector.component.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { IComponentOptions, IController, module } from 'angular';
+
+class ImageSourceSelectorCtrl implements IController {
```
Comment (danielpeach) [inline app/scripts/modules/core/src/artifact/imageSourceSelector.component.ts]: You can omit this controller entirely.

File: app/scripts/modules/core/src/pipeline/config/stages/deploy/deployStage.js [hunk 5]
```diff
@@ -33,6 +33,7 @@ module.exports = angular.module('spinnaker.core.pipeline.stage.deployStage', [
             }
             return stage.clusters.every(cluster =>
               cloudProviderRegistryProvider.$get().getValue(cluster.provider, 'serverGroup.skipUpstreamStageCheck')
+              || cluster.imageSource === 'artifact'
```
Comment (danielpeach) [inline app/scripts/modules/core/src/pipeline/config/stages/deploy/deployStage.js]: This feels a little tacked on - might be better to include a flag in the the provider's config (like `skipUpstreamStageCheck`).

File: app/scripts/modules/core/src/serverGroup/configure/common/deployInitializer.component.ts [hunk 6]
```diff
@@ -77,13 +77,17 @@ export class DeployInitializerController implements IController {
...
     viewState.disableImageSelection = true;
+    viewState.showImageSourceSelector = true;
     viewState.disableStrategySelection = baseCommand.viewState.disableStrategySelection || false;
+    viewState.expectedArtifacts = baseCommand.viewState.expectedArtifacts || [];
     viewState.imageId = null;
     viewState.readOnlyFields = baseCommand.viewState.readOnlyFields || {};
     viewState.submitButtonLabel = 'Add';
     viewState.hideClusterNamePreview = baseCommand.viewState.hideClusterNamePreview || false;
     viewState.templatingEnabled = true;
+
```
Comment (danielpeach) [inline app/scripts/modules/core/src/serverGroup/configure/common/deployInitializer.component.ts]: Extraneous space.

**Comment by ezimanyi** [comment]
@danielpeach : Addressed your comments; trivial changes in 'Next round of PR comments' and pulling out the validation into a function in 'Refactor deploy stage validation into a function'.  Let me know if you have any comments, particularly on the second one.

File: app/scripts/modules/core/src/pipeline/config/stages/deploy/deployStage.js [hunk 7]
```diff
@@ -9,8 +10,9 @@ const angular = require('angular');
 module.exports = angular.module('spinnaker.core.pipeline.stage.deployStage', [
   CLOUD_PROVIDER_REGISTRY,
   SERVER_GROUP_COMMAND_BUILDER_SERVICE,
+  CLUSTER_SERVICE,
 ])
-  .config(function (pipelineConfigProvider, cloudProviderRegistryProvider) {
+  .config(function ($injector, pipelineConfigProvider, cloudProviderRegistryProvider, clusterServiceProvider) {
```
Comment (danielpeach) [inline app/scripts/modules/core/src/pipeline/config/stages/deploy/deployStage.js]: Are you using `$injector` for anything?

File: app/scripts/modules/core/src/cluster/cluster.service.ts [hunk 8]
```diff
@@ -176,6 +176,10 @@ export class ClusterService {
     });
   }
 
+  public isDeployingArtifact(cluster: ICluster): boolean {
+    return (cluster.imageSource === 'artifact');
```
Comment (danielpeach) [inline app/scripts/modules/core/src/cluster/cluster.service.ts]: The parens are valid JS here but not really idiomatic (you'll only see it if you're using React).

File: app/scripts/modules/core/src/pipeline/config/stages/deploy/deployStage.js (hunk 7 above)
Comment (ezimanyi) [reply app/scripts/modules/core/src/pipeline/config/stages/deploy/deployStage.js]: Nope, that's left over from when I was mis-understanding how to get access to the cluster service in the config function; I'll remove it.

File: app/scripts/modules/kubernetes/src/v2/pipelines/stages/deployManifest/deployManifestConfig.html [hunk 9]
```diff
@@ -37,14 +37,16 @@ <h4>Manifest Configuration</h4>
...
 
-  <expected-artifact-selector ng-if="ctrl.$scope.stage.source === ctrl.artifactSource"
-                              command="ctrl.$scope.stage"
-                              id-field="manifestArtifactId"
-                              account-field="manifestArtifactAccount"
-                              accounts="ctrl.metadata.backingData.artifactAccounts"
-                              expected-artifacts="ctrl.expectedArtifacts"
-                              help-field-key="kubernetes.manifest.expectedArtifact">
-  </expected-artifact-selector>
+  <div class="container-fluid form-horizontal">
```
Comment (sbwsg) [inline app/scripts/modules/kubernetes/src/v2/pipelines/stages/deployManifest/deployManifestConfig.html]: we'll end up with an empty `<div>` in the DOM if `ctrl.$scope.stage.source !== ctrl.artifactSource`.  Is this intentional or should the `ng-if` from `expected-artifact-selector` move up to this div?

Comment (ezimanyi) [reply app/scripts/modules/kubernetes/src/v2/pipelines/stages/deployManifest/deployManifestConfig.html]: Good point, I'll fix that

**Comment by ezimanyi** [comment]
Screenshot of the new field in the "Basic Settings" section of "Configure Deployment Cluster"

image

**Pull request merged**